api/v1/composition.go

```diff
@@ -153,24 +154,40 @@ func (c *Composition) InputsExist(syn *Synthesizer) bool {
 		refs[ref.Key] = struct{}{}
 	}
 
+	bound := map[string]struct{}{}
 	for _, binding := range c.Spec.Bindings {
 		if _, ok := refs[binding.Key]; !ok {
 			// Ignore missing resources if the synthesizer doesn't require them
 			// This is important for forwards compatibility- compositions can bind to refs that don't exist, but will in future synths
 			continue
 		}
+		found := slices.ContainsFunc(c.Status.InputRevisions, func(rev InputRevisions) bool {
+			return binding.Key == rev.Key
+		})
+		if !found {
+			return false
+		}
+		bound[binding.Key] = struct{}{}
+	}
 
-		var found bool
-		for _, rev := range c.Status.InputRevisions {
-			if binding.Key == rev.Key {
-				found = true
-				break
+	for _, ref := range syn.Spec.Refs {
+		// Handle missing resources for implied bindings
+		if ref.Resource.Name != "" {
+			found := slices.ContainsFunc(c.Status.InputRevisions, func(rev InputRevisions) bool {
+				return ref.Key == rev.Key
+			})
+			if !found {
+				return false
 			}
+			continue
 		}
-		if !found {
+
+		// Every ref must be bound
```

Comment:
This fixes an unrelated bug: refs without bindings should count as missing inputs